feat: client scene synchronization mode

NoelStephensUnity opened the pull request:
This is a very minor tweak to how clients are synchronized in order to support two different styles of scene loading during client synchronization.  There is a new method, NetworkSceneManager.SetClientSynchronizationMode, that provides users with the ability to configure this feature.
 
**When the ClientSynchronizationMode is set to LoadSceneMode.Single:**
If the server's currently active scene is not already loaded on the client then it will be loaded in single mode. This means that all currently loaded scenes on the client will be unloaded.

**When the ClientSynchronizationMode is set to LoadSceneMode.Additive:**
All currently loaded scenes are left as they are on the client and any newly loaded scenes will be loaded additively.  Users control which scenes are valid for the server to instruct the client to load via the <see cref="VerifySceneBeforeLoading"/> method.

This update also includes the ability to squelch the VerifySceneBeforeLoading console messages which would commonly be used when ClientSynchronizationMode is set to LoadSceneMode.Additive.

The default ClientSynchronizationMode is LoadSceneMode.Single which behaves like the traditional MLAPI initial client synchronization loading pattern.

NoelStephensUnity committed:
refactor

This allows for two different types of client synchronization modes (ClientSynchronizationMode):
LoadSceneMode.Single: All currently loaded scenes on the client will be unloaded and the server's currently active scene will be loaded in single mode on the client unless it was already loaded.

LoadSceneMode.Additive:  All currently loaded scenes are left as they are and any newly loaded scenes will be loaded additively.  Uses need to determine which scenes are valid to load via the VerifySceneBeforeLoading method.

NoelStephensUnity committed:
refactor and style

Added the ability to squelch the validation warning messages for ClientSynchronizationMode additive as it will throw warnings for every loaded scene you don't want to have the client synchronize.

Updated the comments.

NoelStephensUnity committed:
style

fixing spelling typo.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -205,6 +205,22 @@ public class NetworkSceneManager : IDisposable
+        /// scenes will be loaded additively.  Users need to determine which scenes are valid to load via the
+        /// <see cref="VerifySceneBeforeLoading"/> method.
+        /// </summary>
+        public LoadSceneMode ClientSynchronizationMode { get; internal set; }
```
tests for Single & Additive to validate this works in both modes and gives us the results we expect on the client-side? 👀

NoelStephensUnity replied:
Well, for single mode it works already otherwise it wouldn't have passed the already existing tests.
I will add some additional passes in the existing tests which use additive ClientSynchronizationMode.

NoelStephensUnity replied:
Added to the scene loading tests.  It now runs the tests with client synchronization set to both LoadSceneMode.Additive and LoadSceneMode.Single.

NoelStephensUnity committed:
test

added Values to the scene loading test that will run the scene loading tests with client synchronization set to both LoadSceneMode.Single and LoadSceneMode.Additive.

NoelStephensUnity committed:
Merge branch 'develop' into refactor/client-scene-synchronization-mode

0xFA11 commented on testproject/Assets/Tests/Runtime/NetworkSceneManagerTests.cs:
```diff
@@ -43,7 +43,7 @@ private class SceneTestInfo
 
         [UnityTest]
-        public IEnumerator SceneLoadingAndNotifications()
+        public IEnumerator SceneLoadingAndNotifications([Values(LoadSceneMode.Single,LoadSceneMode.Additive)] LoadSceneMode clientSynchronizationMode)
```
so how we verify whether single or additive modes result into single or additive on the client-side?
I'm not seeing any checks or asserts — does this mean that "works fine" implicitly means "unit-tested"?

NoelStephensUnity replied:
Good point, will add verification that the SceneEventData on the client side reflects the clientSynchronization mode set.

NoelStephensUnity committed:
test

Adding verification that the client synchronization mode used by the server when building the client synchronization SceneEventData message is the same value on the client side.

0xFA11 reviewed (approved):
lgtm

NoelStephensUnity committed:
style

fixing whitespace issue
